### PR TITLE
Refactor data/ and mafia.js to for...of

### PR DESCRIPTION
**Pull request by KrisXV**

### CODE REVIEW

File: data/moves.js [hunk 1]
```diff
@@ -5531,18 +5526,17 @@ exports.BattleMovedex = {
...
 						// This move affects every Grass-type Pokemon in play.
-						targets.push(sideActive[activeSlot]);
+						targets.push(activeSlot);
 					}
 				}
 			}
 			let success = false;
-			for (const target of targets) {
-				success = this.boost({def: 1}, target, source, this.getMove('Flower Shield')) || success;
+			for (const pokemon of targets) {
```
Comment (KrisXV) [inline data/moves.js]: i dont remember doing this but im already in bed so ill fix it tomorrow, feel free to point out anything else like this

**Comment by urkerab** [comment]
>     let allActives = pokemon.side.active.concat(pokemon.side.foe.active);

This happens twice but most places seem to use a double loop. Which is preferable?

**Comment by Zarel** [comment]
Double loop is preferable. `allActives` happens in certain cases when we actually need to manipulate/return the array for some reason.

File: data/moves.js [hunk 2]
```diff
@@ -5531,12 +5526,11 @@ exports.BattleMovedex = {
 		flags: {distance: 1},
 		onHitField: function (target, source) {
 			let targets = [];
-			for (let sideSlot = 0; sideSlot < this.sides.length; sideSlot++) {
-				let sideActive = this.sides[sideSlot].active;
-				for (let activeSlot = 0; activeSlot < sideActive.length; activeSlot++) {
-					if (sideActive[activeSlot] && sideActive[activeSlot].isActive && sideActive[activeSlot].hasType('Grass')) {
+			for (const sideSlot of this.sides) {
```
Comment (Zarel) [inline data/moves.js]: `side` not `sideSlot`. `side` is a Side, `sideSlot` is a number.

Comment (Zarel) [reply data/moves.js]: Ditto many of your other uses of "slot".

**Comment by HoeenCoder** [comment]
mafia.js looks good.

**Comment by KrisXV** [comment]
done @Zarel 
@HoeenCoder I made another change (https://github.com/Zarel/Pokemon-Showdown/pull/4490/files#diff-b5323b00c0067cd08016aebb74278873R790) just so you know

**Comment by KrisXV** [comment]
i was aware of that before i pushed too early

**Comment by HoeenCoder** [comment]
Still good, thanks for the heads up!

**Comment by KrisXV** [comment]
Bump?

File: chat-plugins/mafia.js [hunk 3]
```diff
@@ -229,10 +228,10 @@ class MafiaTracker extends Rooms.RoomGame {
 				}
 			}
 			// Determine the role's alignment
-			for (let j = 0; j < target.length; j++) {
-				let id = toId(target[j]);
+			for (let [j, targetId] of target.entries()) {
+				let id = toId(targetId);
 				if (MafiaData.alignments[id]) {
-					if (typeof MafiaData.alignments[id] === 'string') target[j] = MafiaData.alignments[id];
+					if (typeof MafiaData.alignments[id] === 'string') targetId = MafiaData.alignments[id];
```
Comment (Zarel) [inline chat-plugins/mafia.js]: Does this actually do what you think it does?

File: data/abilities.js [hunk 4]
```diff
@@ -100,8 +100,7 @@ exports.BattleAbilities = {
 		onBasePowerPriority: 8,
 		onBasePower: function (basePower, pokemon) {
 			let boosted = true;
-			let allActives = pokemon.side.active.concat(pokemon.side.foe.active);
-			for (const target of allActives) {
+			for (const target of pokemon.side.active.concat(pokemon.side.foe.active)) {
```
Comment (Zarel) [inline data/abilities.js]: I would definitely either do a double `for`-loop, or I would at least define `allActives` on its own line; it's too long for me to like it being inline.

File: data/abilities.js [hunk 5]
```diff
@@ -1130,11 +1125,10 @@ exports.BattleAbilities = {
 	"frisk": {
 		shortDesc: "On switch-in, this Pokemon identifies the held items of all opposing Pokemon.",
 		onStart: function (pokemon) {
-			let foeactive = pokemon.side.foe.active;
-			for (let i = 0; i < foeactive.length; i++) {
-				if (!foeactive[i] || foeactive[i].fainted) continue;
-				if (foeactive[i].item) {
-					this.add('-item', foeactive[i], foeactive[i].getItem().name, '[from] ability: Frisk', '[of] ' + pokemon, '[identify]');
+			for (const foeactive of pokemon.side.foe.active) {
```
Comment (Zarel) [inline data/abilities.js]: `foeActive` would at least match our capitalization rules, but I'd actually call this `target`.

File: data/rulesets.js [hunk 6]
```diff
@@ -783,11 +776,13 @@ exports.BattleFormats = {
...
 				if (template.prevo) types = types.concat(dex.getTemplate(dex.getTemplate(template.prevo).prevo || template.prevo).types);
-				for (let i in baseTemplate.otherFormes) {
-					let forme = dex.getTemplate(baseTemplate.otherFormes[i]);
-					if (baseTemplate.otherFormes && !forme.battleOnly) {
-						if (forme.forme !== 'Alola' && forme.forme !== 'Alola-Totem' && forme.baseSpecies !== 'Wormadam') {
-							types = types.concat(forme.types).concat(baseTemplate.types);
+				if (baseTemplate.otherFormes) {
+					for (const formeid of baseTemplate.otherFormes) {
+						let forme = dex.getTemplate(formeid);
+						if (baseTemplate.otherFormes && !forme.battleOnly) {
```
Comment (Zarel) [inline data/rulesets.js]: The `baseTemplate.otherFormes &&` part seems redundant.

File: data/rulesets.js [hunk 7]
```diff
@@ -717,16 +711,16 @@ exports.BattleFormats = {
 		},
 		onValidateTeam: function (team) {
 			let typeTable;
-			for (let i = 0; i < team.length; i++) {
-				let template = this.getTemplate(team[i].species);
-				if (!template.types) return [`Invalid pokemon ${team[i].name || team[i].species}`];
+			for (const [i, set] of team.entries()) {
```
Comment (Zarel) [inline data/rulesets.js]: `entries()` seems unnecessary here.

File: data/random-teams.js [hunk 8]
```diff
@@ -410,8 +410,8 @@ class RandomTeams extends Dex.ModdedDex {
 		];
 
 		// Iterate through all moves we've chosen so far and keep track of what they do:
-		for (let k = 0; k < moves.length; k++) {
-			let move = this.getMove(moves[k]);
+		for (const [k, moveId] of moves.entries()) {
```
Comment (Zarel) [inline data/random-teams.js]: why `entries()`? is `k` actually used here?

Comment (KrisXV) [reply data/random-teams.js]: yeah, it's really long so you cant find it easily but near the end of the loop it calls for ``moves.splice(k, 1);``

File: data/rulesets.js (hunk 6 above)
Comment (KrisXV) [reply data/rulesets.js]: I forgot to remove that when i moved the if statement out of the loop, oops

File: chat-plugins/mafia.js (hunk 3 above)
Comment (KrisXV) [reply chat-plugins/mafia.js]: yes

File: data/rulesets.js (hunk 7 above)
Comment (KrisXV) [reply data/rulesets.js]: https://github.com/Zarel/Pokemon-Showdown/pull/4490/files#diff-791169e8e004f1c472a40eb12906ea15R717

I mean i'm fine with changing this to `team.indexOf(set)`, it's just that from previous PRs i had developed the assumption that `.entries()` was favorable

File: data/random-teams.js (hunk 8 above)
Comment (KrisXV) [reply data/random-teams.js]: I should probably mention i'm fine with removing `.entries()` and adding `moves.indexOf(moveId)` where k is used independently, but as stated before, precedent

**Comment by KrisXV** [comment]
done @Zarel

File: chat-plugins/mafia.js (hunk 3 above)
Comment (Zarel) [reply chat-plugins/mafia.js]: ...what do you think it does?

As far as I can tell, `targetId` is never read, except in line 232 right after it's declared.

File: data/rulesets.js [hunk 9]
```diff
@@ -717,16 +711,16 @@ exports.BattleFormats = {
 		},
 		onValidateTeam: function (team) {
 			let typeTable;
-			for (let i = 0; i < team.length; i++) {
-				let template = this.getTemplate(team[i].species);
-				if (!template.types) return [`Invalid pokemon ${team[i].name || team[i].species}`];
-				if (i === 0) {
+			for (const set of team) {
```
Comment (Zarel) [inline data/rulesets.js]: Yes, `.entries()` is favorable, I just didn't realize you were still using `i` here... change it back?

File: chat-plugins/mafia.js (hunk 3 above)
Comment (Zarel) [reply chat-plugins/mafia.js]: Poke

Comment (KrisXV) [reply chat-plugins/mafia.js]: I misinterpreted what you asked, whoops, will fix

Comment (KrisXV) [reply chat-plugins/mafia.js]: fixed this and the PR is updated

**Comment by sirDonovan** [comment]
@KrisXV can you rebase this when you get a chance?

**Comment by KrisXV** [comment]
@sirdonovan I can’t do it until I get home in about 3 hours, but I’ll do it asap

**Comment by KrisXV** [comment]
done @Zarel @sirDonovan

**Comment by Zarel** [comment]
Yay! Thanks

**Pull request merged**